### PR TITLE
Show middle state badges for running, starred and checked sub-rows

### PR DESCRIPTION
**Pull request by wolmir**
This PR will introduce indicators for the state of sub-rows in the experiments table.

There will an indicator for selected, starred and running sub-rows.

See https://github.com/iterative/vscode-dvc/issues/1714 for context on the issue.

https://user-images.githubusercontent.com/1231848/179523374-1acbe6c8-d3e4-43c0-8c5e-6e82aecb8334.mov

### CODE REVIEW

File: webview/src/shared/components/tooltip/Tooltip.tsx [hunk 1]
```diff
@@ -34,7 +35,7 @@ const TooltipRenderFunction: React.ForwardRefRenderFunction<
 ) => (
   <Tippy
     animation={animation}
-    appendTo={'parent'}
+    appendTo={appendTo}
```
Comment (wolmir) [inline webview/src/shared/components/tooltip/Tooltip.tsx:38]: Thanks @rogermparent !

**Comment by mattseddon** [comment]
[Q] Why are we not showing the middle state for experiments that have been selected for plotting?

File: webview/src/experiments/components/table/Cell.tsx [hunk 2]
```diff
@@ -31,25 +33,111 @@ const RowExpansionButton: React.FC<RowProp> = ({ row }) =>
...
+        tooltipContent={
+          showSubRowStates &&
+          subRowsAffected &&
+          `${subRowsAffected} ${pluralize(
+            'sub-row',
+            subRowsAffected
+          )} ${actionAppliedLabel}.`
+        }
+      >
+        {children}
```
Comment (mattseddon) [inline webview/src/experiments/components/table/Cell.tsx:76]: [I] At the moment only the direct children are being aggregated. Here is a screen recording of an example:

https://user-images.githubusercontent.com/37993418/178470947-92586894-b9e2-45c3-b1cb-2a990f0b7667.mov

When we aggregate from checkpoints to experiments and then up to the current commit we are not taking the cumulative total into account.

Another example:

https://user-images.githubusercontent.com/37993418/178471450-fa337248-913a-43d9-8175-cac4a44be943.mov

I would not even try to include the top level indicator until we have more than one commit/branch in the table. Until that time we will be catering to an edge case.

However, if you feel that they are needed then they need to reflect the totals including all children.

File: webview/src/experiments/components/table/styles.module.scss [hunk 3]
```diff
@@ -450,19 +454,58 @@ $workspace-row-edge-margin: $edge-padding - $cell-padding;
...
+        }
+
+        .bullet {
+          flex: 0 0 14px;
+          width: 14px;
+          &::after {
+            border: none;
+          }
+
+          &.subRunsCount {
```
Comment (mattseddon) [inline webview/src/experiments/components/table/styles.module.scss:496]: If we are going to push the contents of the first column right to make space for this indicator then we need to increase the min width of the first column:

image

Comment (mattseddon) [reply webview/src/experiments/components/table/styles.module.scss:496]: picture is from Storybook snapshot (names are completely hidden)

File: webview/src/experiments/components/table/Cell.tsx (hunk 2 above)
Comment (wolmir) [reply webview/src/experiments/components/table/Cell.tsx:76]: I'll add some tests for this current use-case now, but I want to defer this total aggregation for later if it's currently an edge case.

**Comment by wolmir** [comment]
I just added the plots indicator and will add some tests for all of them now

**Comment by wolmir** [comment]
> This pull request will bring the total coverage in the repository to 96.6% (0.0% change).

It was 96.7% **before** I added more tests. Go figure

File: webview/src/experiments/components/App.test.tsx [hunk 4]
```diff
@@ -372,6 +377,213 @@ describe('App', () => {
...
+        const indicator = getCountIndicatorById(row, 'row-subrows-running')
+        expect(indicator).not.toBeInTheDocument()
+      })
+
+      it('should display the correct number of running sub-rows when the parent is collapsed', () => {
+        renderWithAllRowsExpanded()
+        const row = getMiddleStatesTestRow()
+        const numberOfSubrowsRunning = runSomeExperiments()
+        expect(
+          getCountIndicatorById(row, 'row-subrows-running')
```
Comment (mattseddon) [inline webview/src/experiments/components/App.test.tsx:553]: [I] Seeing as we are no longer showing this information for the HEAD commit we no longer need the sub row running indicator. We should remove all of the code and shift the chevron, etc in the first cell to the left (saves space).

File: webview/src/experiments/components/App.test.tsx [hunk 5]
```diff
@@ -372,6 +377,213 @@ describe('App', () => {
...
+        toggleExpansion(row)
+        const collapsed = getMiddleStatesTestRow()
+        expect(
+          getCountIndicatorById(collapsed, 'row-subrows-running')
+        ).toHaveTextContent(`${numberOfSubrowsRunning}`)
+      })
+    })
+
+    describe('Plots counter', () => {
+      it('should not be visible if no sub-row experiment is selected', () => {
```
Comment (mattseddon) [inline webview/src/experiments/components/App.test.tsx:564]: [C] As discussed we want this on top of the plots indicator, not separate (again, save space).

File: webview/src/experiments/components/App.test.tsx [hunk 6]
```diff
@@ -372,6 +377,213 @@ describe('App', () => {
...
+        renderWithAllRowsExpanded()
+        const row = getMiddleStatesTestRow()
+        const indicator = getCountIndicatorById(row, 'row-subrows-selected')
+        expect(indicator).not.toBeInTheDocument()
+      })
+
+      it('should display the correct number of running sub-rows when the parent is collapsed', () => {
```
Comment (mattseddon) [inline webview/src/experiments/components/App.test.tsx:571]: [I] Update text for plots selection (match description).

File: webview/src/test/tableDataFixture.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    return colors.pop()
+  }
+
+  return transformRows(fixture, labelOrIds, row => ({
+    ...row,
+    displayColor: nextColor(),
+    selected: true
+  }))
+}
+export const setExperimentsAsRunning = setRowPropertyAsTrue('running')
```
Comment (wolmir) [inline webview/src/test/tableDataFixture.ts:69]: I added these helpers to remove some of the friction I had with adding new tests. Hopefully this will make it way easier to setup table data scenarios for new tests.

File: webview/src/experiments/components/table/styles.module.scss [hunk 8]
```diff
@@ -450,19 +454,57 @@ $workspace-row-edge-margin: $edge-padding - $cell-padding;
...
+        }
+
+        .indicatorCount {
+          z-index: 2;
+          &[title='0'] {
+            display: none;
+          }
+        }
+
+        .bullet {
```
Comment (wolmir) [inline webview/src/experiments/components/table/styles.module.scss:489]: I need to remove this, they are not used anymore

File: webview/src/experiments/components/table/styles.module.scss [hunk 9]
```diff
@@ -672,3 +714,26 @@ $badge-size: 0.85rem;
 .cellTooltip {
   padding: 2px 6px;
 }
+
+.indicatorCount {
+  &[aria-label='0'] {
```
Comment (wolmir) [inline webview/src/experiments/components/table/styles.module.scss:719]: I tried using only this style to not show the counter when the number is 0, because it would save a conditional test in the component and reduce its complexity.

It works on VSCode and Storybook, but for some reason these styles are not applied when running the jest tests.
So I kept the conditionals in the end.

File: webview/src/experiments/components/App.test.tsx [hunk 10]
```diff
@@ -372,6 +373,129 @@ describe('App', () => {
     })
   })
 
+  describe('Sub-rows middle states indicators', () => {
+    const renderWithAllRowsExpanded = () => {
```
[the remark this answers is not in the document]
Comment (mattseddon) [reply webview/src/experiments/components/App.test.tsx:377]: Splitting out utils will help to start reducing the size of this file and would give a central location for people to look for test building blocks.

[the remark this answers is not in the document]
Comment (mattseddon) [reply webview/src/experiments/components/App.test.tsx:377]: fine to create a separate PR for that

File: webview/src/experiments/components/table/Indicators.tsx [hunk 11]
```diff
@@ -3,51 +3,97 @@ import { SortDefinition } from 'dvc/src/experiments/model/sortBy'
...
+}
+
+export const CounterBadge: React.FC<CounterBadgeProps> = ({ count }) => {
+  return count ? (
+    <span
+      className={styles.indicatorCount}
+      role={'marquee'}
+      aria-label={`${count ?? 0}`}
+    >
+      {count ?? 0}
```
Comment (mattseddon) [inline webview/src/experiments/components/table/Indicators.tsx:49]: [Q] How is `??` needed here when it can only be a truthy value?. Could you use a [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) here?

File: webview/src/experiments/components/table/styles.module.scss (hunk 8 above)
Comment (mattseddon) [reply webview/src/experiments/components/table/styles.module.scss:489]: Please remove this before merging

File: webview/src/stories/Table.stories.tsx [hunk 12]
```diff
@@ -35,9 +39,11 @@ const tableData: TableData = {
     ...row,
     subRows: row.subRows?.map(experiment => ({
       ...experiment,
+      starred: experiment.starred || experiment.label === '42b8736',
       subRows: experiment.subRows?.map(checkpoint => ({
         ...checkpoint,
-        running: checkpoint.running || checkpoint.label === '23250b3'
+        running: checkpoint.running || checkpoint.label === '23250b3',
+        starred: checkpoint.starred || checkpoint.label === '22e40e1'
```
[the remark this answers is not in the document]
Comment (mattseddon) [reply webview/src/stories/Table.stories.tsx:46]: 👍🏻

**Review by mattseddon — APPROVED** [review]
Just one thing that I would recommend doing before you merge this. Adding an updated video in description would be handy for future reference too.
